### ISSUE TITLE
MonoNext.NextSubscriber.onNext() calls s.cancel() before actual.onNext(t).This can lead to unexpected Thread interrupt

### ISSUE THREAD

**Issue by ab-taktik** [reactor/reactor-core#962]
Hello, 

I have a MonoPublishOnSubscriber that, from its run() method, ultimately calls a MonoNext.NextSubscriber.onNext().

MonoNext.NextSubscriber.onNext() has the following code :

```java
...
    s.cancel();
    actual.onNext(t);
    onComplete();
...
```

So `s.cancel()` is called **before** ` actual.onNext(t);`

When `s.cancel();` is called, it goes back to the `MonoPublishOnSubscriber.cancel()`method that calls dispose(), effectively interrupting the current thread.

Then, when `actual.onNext(t);` is called, I got a `java.lang.InterruptedException` because some library code in the callback is checking the current thread interrupt flag.

To be complete, this only happens if I publish on an elastic scheduler, not a parallel one, which adds to the confusion. 

My question is, is this expected behaviour? And if so, is there still a way to use an elastic scheduler without interrupting on cancel()?

Thanks in advance, 

Antoine

Edit : 
Looking at the code, it would be nice to specify the "interrupt on cancel" behaviour as a parameter of the ElasticScheduler. Currently this is hardcoded : 

```java
	@Override
	public Disposable schedule(Runnable task) {
		ScheduledExecutorService exec = pick();

		Runnable wrapper = () -> {
			try {
				task.run();
			}
			catch (Throwable ex) {
				Schedulers.handleError(ex);
			}
			finally {
				release(exec);
			}
		};
		Future<?> f;

		try {
			f = exec.submit(wrapper);
		}
		catch (RejectedExecutionException ex) {
			return REJECTED;
		}
		return new ExecutorServiceScheduler.DisposableFuture(f, true); // Hardcoded value
	}
```

**Comment by simonbasle** [comment]
which version of reactor-core is that? Elastic has changed in 3.1...

The cancel before onNext is on purpose, preventing a chatty source to hog the thread when you'd only be interested in the first value.

Can you test with the latest version (I know, if you're on 3.0.x you'll need some migration effort)? Can you provide a self-contained reproduction unit test?

**Comment by ab-taktik** [comment]
Sorry, it's 3.0.7
I'll try to upgrade and let you know

**Comment by ab-taktik** [comment]
Can confirm that this is fixed in 3.1 : There is a check in SchedulerTask.dispose() : 

```java
f.cancel(thread != Thread.currentThread());
```
which evaluates to "false" in my case, so no interrupt is triggered.

Thanks for your help

**Comment by simonbasle** [comment]
great 👍
